_blog/2023-12-15-using-actix-rust.mdx

```diff
@@ -0,0 +1,434 @@
+
+// This future doesn't have the requirement of being `Send`.
+// See: futures_util::future::LocalBoxFuture
+type LocalBoxFuture<T> = Pin<Box<dyn Future<Output = T> + 'static>>;
+
+// `S`: type of the wrapped service
+// `B`: type of the body - try to be generic over the body where possible
+impl<S, B> Service<ServiceRequest> for SayHiMiddleware<S>
+where
+    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
+    S::Future: 'static,
+    B: 'static,
+{
+    type Response = ServiceResponse<B>;
+    type Error = Error;
+    type Future = LocalBoxFuture<Result<Self::Response, Self::Error>>;
+
+    // This service is ready when its next service is ready
+    forward_ready!(service);
+
+    fn call(&self, req: ServiceRequest) -> Self::Future {
+        println!("Hi from start. You requested: {}", req.path());
+
+        // A more complex middleware, could return an error or an early response here.
+
+        // we do not immediately await this, which means nothing happens
+        // this future gets moved into a Box
+        let fut = self.service.call(req);
+
+        Box::pin(async move {
+            // this future gets awaited now
+            let res = fut.await?;
+
+            // we can now do any work we need to after the request
+            println!("Hi from response");
+            Ok(res)
+        })
+    }
+}
+```
+Now that we've written our middleware, we can now add it to our App:
+```rust
+#[actix_web::main]
+async fn main() -> std::io::Result<()> {
+// access logs are printed with the INFO level so ensure it is enabled by default
+env_logger::init_from_env(env_logger::Env::new().default_filter_or("info"));
+
+let app = App::new()
+     .wrap(SayHi);
+     // ... rest of your code
+}
+```
+
+## Static Files in Actix Web
+Plain, no-frills static file serving in [`actix-web`](https://actix.rs/) is done through the `actix_files` crate - to add it, you simply need to add it through Cargo like so:
+```bash
+cargo add actix-files 
+```
+
+Setting up a route for static file serving would look like this:
+```rust
+use actix_files::NamedFile;
+use actix_web::{HttpRequest, Result};
+use std::path::PathBuf;
+
+async fn index(req: HttpRequest) -> Result<NamedFile> {
+    let path: PathBuf = req.match_info().query("filename").parse().unwrap();
+    Ok(NamedFile::open(path)?)
+}
+```
+This route allows us to serve any file that can be found and matches the filename - for example, if we have a base route that serves this route, if we then run our app and go to `/index.html`, the route will try to look for a file named `index.html` in the project root.
+
+However, this is a bit clumsy: let's have a look at writing a route that can search for any HTML file in a directory where we may have stored some HTML files.
+```rust
+#[get("/{filename}")] // <- define path parameters
+async fn index(filename: String) -> Result<NamedFile> {
+    let filename = format!("dir_name/{filename}.html");
+    Ok(NamedFile::open(path)?)
+}
+```
```

Comment:
Is this really the way to do it? Both of these look vulnerable to path traversal

Reply:
That's the way it's shown in the docs, but it appears from their GitHub issues/PRs that this was a known issue and is documented here: https://actix.rs/docs/static-files

I guess we should probably insert some kind of warning here to make sure readers know to not try and end their paths in `.*`

Reply:
Yeah add some little warning. Not sure if the current non-tail match is safe or not.